botocore/client.py

```diff
@@ -325,6 +246,144 @@ def _api_call(self, *args, **kwargs):
         return _api_call
 
 
+class ClientEndpointBridge(object):
+    """Bridges endpoint data and client creation
+
+    This class handles taking out the relevant arguments from the endpoint
+    resolver and determining which values to use, taking into account any
+    client configuration options and scope configuration options.
+
+    This class also handles determining what, if any, region to use if no
+    explicit region setting is provided. For example, Amazon S3 client will
+    utilize "us-east-1" by default if no region can be resolved."""
+
+    SIGNATURE_PRECEDENCE = ['s3', 's3v4', 'v4', 'v2']
+    DEFAULT_ENDPOINT = '{service}.{region}.amazonaws.com'
+
+    def __init__(self, endpoint_resolver, scoped_config,
+                 client_config, default_endpoint=None):
+        self.endpoint_resolver = endpoint_resolver
+        self.scoped_config = scoped_config
+        self.client_config = client_config
+        self.default_endpoint = default_endpoint or self.DEFAULT_ENDPOINT
+
+    def resolve(self, service_name, region_name=None, endpoint_url=None,
+                is_secure=True):
+        region_name = self._check_default_region(service_name, region_name)
+        resolved = self.endpoint_resolver.construct_endpoint(
+            service_name, region_name)
+        if resolved:
+            return self._create_endpoint(
+                resolved, service_name, region_name, endpoint_url, is_secure)
+        else:
+            return self._assume_endpoint(service_name, region_name,
+                                         endpoint_url, is_secure)
+
+    def _check_default_region(self, service_name, region_name):
+        if region_name is not None:
+            return region_name
+        # Use the client_config region if no explicit region was provided.
+        if self.client_config and self.client_config.region_name is not None:
+            return self.client_config.region_name
+        # Use us-east-1 as the default region for S3.
+        if service_name == 's3':
+            return 'us-east-1'
```

Comment:
Is there anyway we can avoid hard coded logic of service names/region names in code?  I get this is special behavior we need, but previously this was captured in the old endpoints file.  Is there anyway this information can be moved into endpoints.json?